File: mod/message_stats.py
"""Collect basic statistics on messaging activity per user."""

from datetime import datetime
from storage import *
import discord
from client import client

mod = client.module("MessageStats")


@mod.on_message
@orm.db_session
async def on_message(msg: discord.Message):
    if msg.content != "":
        ensure_user_created(msg.author.id, msg.author.name)
        Message(
            id=msg.id,
            author=msg.author.id,
            length=len(msg.content),
            channel=msg.channel.id,
            time=datetime.now(),
        )
        db.commit()


@client.tree.command()
async def mymessagestats(interaction: discord.Interaction):
    with orm.db_session:
        msg_count = orm.count(m for m in Message if m.author.id == interaction.user.id)
    await interaction.response.send_message(
        f"You have sent {msg_count} message" + ("" if msg_count == 1 else "s"),
        ephemeral=True,
    )

File: src/mod/messageStats.ts
import { ChatInputCommandInteraction, SlashCommandBuilder } from "discord.js";
import { Module, SlashCommand } from "./module";
import db, { ensureUserExists } from "../storage";

const messageStats: Module = {
  name: "MessageStats",
  async setup(client) {
    client.on("messageCreate", async (msg) => {
      if (!msg.inGuild()) return;
      await ensureUserExists(msg.author.id, msg.member?.nickname ?? msg.author.username);
      const row = {
        author: BigInt(msg.author.id),
        length: msg.content.trim().length,
        channel: BigInt(msg.channelId),
        time: new Date(),
      };
      await db.message.upsert({
        create: {
          id: Number(msg.id),
          ...row,
        },
        update: row,
        where: {
          id: Number(msg.id),
        },
      });
    });
  },
  commands: [
    class extends SlashCommand {
      name = "mymessagestats";
      description = "Check basic statistics about the messages I have sent here.";

      build(builder: SlashCommandBuilder) {
        builder.setDMPermission(false);
      }

      async run(interaction: ChatInputCommandInteraction) {
        const count = await db.message.count({
          where: { author: { equals: BigInt(interaction.user.id) } },
        });
        await interaction.reply({
          content: `You have sent ${count} message${count === 1 ? "" : "s"}`,
          ephemeral: true,
        });
      }
    },
  ],
};

export default messageStats;

